pkg/controller/volume/pvcprotection/pvc_protection_controller.go

```diff
@@ -41,6 +42,65 @@ import (
 
 // Controller is controller that removes PVCProtectionFinalizer
 // from PVCs that are used by no pods.
+
+type LazyLivePodList struct {
+	cache      []v1.Pod
+	controller *Controller
+}
+
+func (ll *LazyLivePodList) getCache() []v1.Pod {
+	return ll.cache
+}
+
+func (ll *LazyLivePodList) setCache(pods []v1.Pod) {
+	ll.cache = pods
+}
+
+type pvcData struct {
+	pvcKey  string
+	pvcName string
+}
+
+type pvcProcessingStore struct {
+	namespaceToPVCsMap map[string][]pvcData
+	namespaceQueue     workqueue.TypedInterface[string]
+	mu                 sync.Mutex
+}
+
+func NewPVCProcessingStore() *pvcProcessingStore {
+	return &pvcProcessingStore{
+		namespaceToPVCsMap: make(map[string][]pvcData),
+		namespaceQueue:     workqueue.NewTyped[string](),
+	}
+}
+
+func (m *pvcProcessingStore) addOrUpdate(namespace string, pvcKey, pvcName string) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	if _, exists := m.namespaceToPVCsMap[namespace]; !exists {
+		m.namespaceToPVCsMap[namespace] = make([]pvcData, 0)
+		m.namespaceQueue.Add(namespace)
+	}
+	m.namespaceToPVCsMap[namespace] = append(m.namespaceToPVCsMap[namespace], pvcData{pvcKey: pvcKey, pvcName: pvcName})
+}
+
+// Returns a list of pvcs and the associated namespace to be processed downstream
+func (m *pvcProcessingStore) flushNextPVCsByNamespace() ([]pvcData, string) {
+
+	nextNamespace, quit := m.namespaceQueue.Get()
+	if quit {
+		return nil, nextNamespace
+	}
+
```

Comment:
I think there might be a possibility of a race condition here that could cause premature consumer goroutine shutdown:
1) Consumer A gets namespace X from the queue
2) Prior to locking `m.mu`, Producer re-adds namespace X to the queue (this is marked as dirty, since X is being processed). This adds a new PVC to `m.namespaceToPVCsMap`
3) Consumer A locks `m.namespaceToPVCsMap`, processes all PVCs for X (including the PVC added in step 2) and deletes namespace X from `m.namespaceToPVCsMap`
4) Consumer B (or consumer A again) enters and gets namespace X from the queue
5) Consumer B looks up X in `m.namespaceToPVCsMap`, gets nil
6) Caller of this method returns false, because pvcList is nil.
7) This causes `runProcessNamespaceWorker` consumer B goroutine to exit.

It might be better to return a boolean, which indicates if the caller should `quit`, rather than relying on values of pvcList and namespace. I think that would make things more clear and less error prone (if there is some other condition that could result in an empty pvcList), that you're only propagating the `quit` signal to the caller, based on the queue shutdown.

Reply:
Apologies, my mistake.

@hungnguyen243 pointed out that (2) will not happen, as the producer does not re-add to the queue (it only appends to the `m.namespaceToPVCsMap` when the lock is acquired), so this prevents this scenario.